pippy/PipelineSchedule.py

```diff
@@ -0,0 +1,460 @@
+import torch.multiprocessing as mp
+import torch.nn as nn
+from torch.profiler import profile, ProfilerActivity, record_function
+
+logger = logging.getLogger(__name__)
+
+
+class PipelineStage(nn.Module):
+    def __init__(
+        self,
+        module: nn.Module,
+        stage_id: int,
+        num_stages: int,
+        rank: int,
+        world_size: int,
+        meta_input: torch.Tensor,
+    ):
+        super().__init__()
+        self.rank = rank
+        self.stage_id = stage_id
+        self.is_first_stage = stage_id == 0
+        self.is_last_stage = stage_id == num_stages - 1
+        self.num_stages = num_stages
+        # When we materialize the model partition on cuda, we call reset_parameters() if it is available
+        self.module = module.to(device=torch.cuda.current_device())
+        if hasattr(self.module, "reset_parameters"):
+            with torch.no_grad():
+                self.module.reset_parameters()
+
+        meta_output = self.module(meta_input)
+        self.fwd_input = torch.empty_like(meta_input, device="cuda")
+        self.fwd_output = None
+        self.fwd_output_grads = torch.empty_like(meta_output, device="cuda")
+        self.fwd_outputs_for_backward = deque()
+
+        self.prev_stage = (rank - 1) % world_size
+        self.next_stage = (rank + 1) % world_size
+
+        self.fwd_recv_queue = None
+        self.bwd_recv_queue = None
+
+        self.requests: List[dist.P2POp] = []
+        logger.info(
+            f"finished pipeline stage init, {self.stage_id=}, {self.is_first_stage=}, {self.is_last_stage=}, {self.num_stages=}, {self.fwd_input.shape=}, {self.fwd_output_grads.shape=}"
+        )
+
+    def init_p2p_neighbors(self):
+        """
+        Set up p2p communitors between previous and next stages
+        by sending a dummy tensor.
+
+        If this is used, must be called for all pipeline stages.
+        """
+        ops = []
+        recv_tensor = torch.zeros(1, device="cuda")
+        send_tensor = torch.ones(1, device="cuda")
+        # forward
+        if not self.is_first_stage:
+            ops.append(dist.P2POp(dist.irecv, recv_tensor, self.prev_stage))
+        if not self.is_last_stage:
+            ops.append(dist.P2POp(dist.isend, send_tensor, self.next_stage))
+
+        # backward
+        if not self.is_first_stage:
+            ops.append(dist.P2POp(dist.isend, send_tensor, self.prev_stage))
+        if not self.is_last_stage:
+            ops.append(dist.P2POp(dist.irecv, recv_tensor, self.next_stage))
+
+        return True
+
+    def get_fwd_recv_ops(self) -> List[dist.P2POp]:
+        if self.is_first_stage:
+            return []
+        return [dist.P2POp(dist.irecv, self.fwd_input, self.prev_stage)]
+
+    def get_fwd_send_ops(self) -> List[dist.P2POp]:
+        if self.is_last_stage:
+            return []
+        return [dist.P2POp(dist.isend, self.fwd_output, self.next_stage)]
+
+    def forward(self, input_data, is_first_mb, is_last_mb):
```

Comment:
`is_first_mb` & `is_last_mb` used only for logging?

Reply:
yep, only used for logging. Before it was used when communication was done within forward, but now communication (batch_isend_irecv) is outside in the schedule implementation so its not longer checked.